graylog2-server/src/main/java/org/graylog2/shared/plugins/PluginLoader.java

```diff
@@ -87,8 +89,9 @@ public PluginLoader(File pluginDir) {
             } else {
                 try {
                     LOG.debug("Loading <" + jar.getAbsolutePath() + ">");
-                    final ClassLoader pluginClassLoader = new URLClassLoader(new URL[]{jar.toURI().toURL()}, classLoader);
-                    final ServiceLoader<Plugin> pluginServiceLoader = ServiceLoader.load(Plugin.class, pluginClassLoader);
+                    final ClassLoader pluginClassLoader = URLClassLoader.newInstance(new URL[]{jar.toURI().toURL()});
+                    classLoader.addClassLoader(pluginClassLoader);
+                    final ServiceLoader<Plugin> pluginServiceLoader = ServiceLoader.load(Plugin.class, classLoader);
```

Comment:
Can this create problems if we ever try to unload plugins?
The ChainingClassLoader owns all plugin classloaders, which would prevent them from being garbage collected, as opposed to the other way around when plugin classloaders get a parent.

Since we need the ability to see classes from other plugins in a plugin as we create more extension points, making the chaining loader accessible is right, but I would prefer to pass it as a parent loader to the URLClassLoader and making the ChainingLoader smarter in not delegating to the calling loader (i.e. to prevent infinite recursion).

That should allow clean unloading if we ever implement it by simply removing a plugin's classloader from the ChainingClassLoader (and additionally cleaning the class usages, of course, but that's out of scope here).